tests/api/enqueue.cpp

```diff
@@ -439,3 +439,147 @@ TEST_F(WithCommandQueue, FinishAfterFlush) {
+    }
+    cl_event event;
+    SetKernelArg(kernel, 1, &i);
+    EnqueueNDRangeKernel(kernel, 1, nullptr, &gws, &lws, 0, nullptr, &event);
+    i++;
+    Flush();
+
+    SetKernelArg(kernel, 1, &i);
+    cl_uint err;
+    cl_uint count = 0;
+    do {
+        err = clEnqueueNDRangeKernel(m_queue, kernel, 1, nullptr, &gws, &lws, 0,
+                                     nullptr, nullptr);
+        count++;
+    } while (err == CL_OUT_OF_RESOURCES);
+    cl_int event_cmd_execution_status;
+    GetEventInfo(event, CL_EVENT_COMMAND_EXECUTION_STATUS,
+                 &event_cmd_execution_status);
+    ASSERT_EQ(event_cmd_execution_status, CL_COMPLETE);
+    ASSERT_GT(count, 10);
+    ASSERT_CL_SUCCESS(clReleaseEvent(event));
+}
+
+TEST_F(WithCommandQueue, EnqueueTooManyCommandsWithRetry) {
+
+    // Make sure that we will have to retry
+    static const unsigned NUM_INSTANCES =
+        CLVK_CONFIG_GET(max_entry_points_instances);
+    static const unsigned NUM_INSTANCES_RETRY = NUM_INSTANCES + 16;
+
+    static const char* program_source = R"(
+    kernel void test_simple(global uint* out, uint id)
+    {
+        out[id] = id;
+    }
+    )";
+
+    auto cfg_force_descriptor_set_allocation_failure =
+        CLVK_CONFIG_SCOPED_OVERRIDE(force_descriptor_set_allocation_failure,
+                                    bool, true, true);
+    auto cfg_descriptor_sets_allocate_retry_sleep_us =
+        CLVK_CONFIG_SCOPED_OVERRIDE(descriptor_sets_allocate_retry_sleep_us,
+                                    uint32_t, 100, true);
+    // We need to make sure that the kernel don't start executing before we
+    // enqueue one too many
+    CLVK_CONFIG_ASSERT_GT(max_first_cmd_batch_size, NUM_INSTANCES);
+
+    // Create kernel
+    auto kernel = CreateKernel(program_source, "test_simple");
+
+    // Create buffer
+    size_t buffer_size = NUM_INSTANCES_RETRY * sizeof(cl_uint);
+    auto buffer = CreateBuffer(CL_MEM_WRITE_ONLY | CL_MEM_ALLOC_HOST_PTR,
+                               buffer_size, nullptr);
+
+    // Dispatch kernel
+    size_t gws = 1;
+    size_t lws = 1;
+
+    SetKernelArg(kernel, 0, buffer);
+    cl_uint i;
+    for (i = 0; i < NUM_INSTANCES; i++) {
+        SetKernelArg(kernel, 1, &i);
+        EnqueueNDRangeKernel(kernel, 1, nullptr, &gws, &lws);
+    }
+    Flush();
+
+    cl_uint err = CL_OUT_OF_RESOURCES;
+    SetKernelArg(kernel, 1, &i);
+    i++;
+
+    // wait for kernel to start
+    auto ts_start = sampleTime();
+    const uint64_t one_second = 1ull * 1000 * 1000 * 1000;
+    while (sampleTime() < ts_start + one_second && err == CL_OUT_OF_RESOURCES) {
+        err = clEnqueueNDRangeKernel(m_queue, kernel, 1, nullptr, &gws, &lws, 0,
+                                     nullptr, nullptr);
+    }
+    ASSERT_CL_SUCCESS(err);
+
+    for (; i < NUM_INSTANCES_RETRY; i++) {
+        SetKernelArg(kernel, 1, &i);
+        EnqueueNDRangeKernel(kernel, 1, nullptr, &gws, &lws);
+    }
```

Comment:
Similar question here: I'm not sure I understand why you're doing the enqueues in three groups or why you're swallowing some enqueue errors. This doesn't seem to be the behaviour that we want/ Are we missing an automatic flush in the runtime when we run out of descriptor sets?

Reply:
This is complicated. You could consider that we are missing an automatic flush, but this is not an easy one to insert. Basically we are failing the descriptor set allocation when we are here:
```
  * frame #0: 0x00007ffff770637e libOpenCL.so.1`cvk_command_kernel::build_batchable_inner(this=0x0000555556086a00, command_buffer=0x0000555556086bb0) at queue.cpp:985:9
    frame #1: 0x00007ffff7706ec4 libOpenCL.so.1`cvk_command_batchable::build(this=0x0000555556086a00, command_buffer=0x0000555556086bb0) at queue.cpp:1169:16
    frame #2: 0x00007ffff770bbe4 libOpenCL.so.1`cvk_command_batch::add_command(this=0x0000555555bb1460, cmd=0x0000555556086a00) at queue.hpp:830:27
    frame #3: 0x00007ffff7700a90 libOpenCL.so.1`cvk_command_queue::enqueue_command(this=0x0000555555b98960, cmd=0x0000555556086a00, event=0x0000000000000000) at queue.cpp:150:32
    frame #4: 0x00007ffff7701854 libOpenCL.so.1`cvk_command_queue::enqueue_command_with_deps(this=0x0000555555b98960, cmd=0x0000555556086a00, num_dep_events=0, dep_events=0x0000000000000000, event=0x0000000000000000) at queue.cpp:209:12
    frame #5: 0x00007ffff76236dd libOpenCL.so.1`cvk_enqueue_ndrange_kernel(command_queue=0x0000555555b98960, kernel=0x0000555555ba9d20, dims=1, ndrange=0x00007fffffffd610, num_events_in_wait_list=0, event_wait_list=0x0000000000000000, event=0x0000000000000000) at api.cpp:4016:27
    frame #6: 0x00007ffff7623f1a libOpenCL.so.1`clEnqueueNDRangeKernel(command_queue=0x0000555555b98978, kernel=0x0000555555ba9d38, work_dim=1, global_work_offset=0x0000000000000000, global_work_size=0x00007fffffffd830, local_work_size=0x00007fffffffd828, num_events_in_wait_list=0, event_wait_list=0x0000000000000000, event=0x0000000000000000) at api.cpp:4075:12
 ```
 This makes it quite hard to flush at that point.
 Maybe we should not wait in build_batchable_inner, and do that in enqueue_command to be able to flush.
 I'll have a look to see if that is possible.